README.md
# overlay_tutorial

A Flutter package for displaying overlay tutorial.

# Feature

- Allow to display an overlay tutorial with hole cropping.
- Allow to display hint aside hole.
- Allow to customize hole shape.

# Usage

Wrap the widget that you want to create a hole using `OverlayTutorialHole`. Ensure you have `OverlayTutorialScope` as a parent of these holes. 

1. Use `OverlayTutorialScope.enabled` property to control the visibility of the overlay tutorial. If this is false, the entire overlay tutorial will be hidden.
2. Use `OverlayTutorialHole.enabled` property to control the visibility of the holes.

## Wrap your UI

```dart

Widget build(BuildContext context) {
  final textTheme = Theme.of(context).textTheme;
  const tutorialColor = Colors.yellow;
  return SafeArea(
    child: Stack(
      children: [
        OverlayTutorialScope(
          enabled: _isTutorialEnabled,
          overlayColor: Colors.blueAccent.withOpacity(.6),
          // Disable all the widgets. All the widgets are now non-interactive.
          child: AbsorbPointer(
            absorbing: _isTutorialEnabled,
            child: Scaffold(
              appBar: AppBar(
                title: Text(widget.title),
              ),
              body: Center(
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    AnimatedBuilder(
                      animation: _tweenAnimation,
                      builder: (context, child) {
                        return SlideTransition(
                          position: _tweenAnimation,
                          child: child,
                        );
                      },
                      child: OverlayTutorialHole(
                        enabled: true,
                        overlayTutorialEntry: OverlayTutorialCustomShapeEntry(
                          shapeBuilder: (rect, path) {
                            path = Path.combine(
                              PathOperation.difference,
                              path,
                              Path()
                                ..addOval(Rect.fromLTWH(
                                  rect.left - 16,
                                  rect.top,
                                  112,
                                  64,
                                )),
                            );
                            return path;
                          },
                        ),
                        child: const Icon(
                          Icons.share,
                          size: 64,
                        ),
                      ),
                    ),
                    const SizedBox(height: 64),
                    const Text(
                      'You have pushed the button this many times:',
                    ),
                    OverlayTutorialHole(
                      enabled: true,
                      overlayTutorialEntry: OverlayTutorialRectEntry(
                        padding: const EdgeInsets.symmetric(horizontal: 8.0),
                        radius: const Radius.circular(16.0),
                        overlayTutorialHints: <OverlayTutorialWidgetHint>[
                          OverlayTutorialWidgetHint(
                            position: (rect) => Offset(0, rect.bottom),
                            builder: (context, entryRect) {
                              return SizedBox(
                                width: MediaQuery.of(context).size.width,
                                child: Center(
                                  child: Padding(
                                    padding: const EdgeInsets.all(8.0),
                                    child: Text(
                                      'Current Counter will be displayed here',
                                      style: textTheme.bodyText2!
                                          .copyWith(color: tutorialColor),
                                    ),
                                  ),
                                ),
                              );
                            },
                          ),
                        ],
                      ),
                      child: Text(
                        '$_counter',
                        style: Theme.of(context).textTheme.headline4,
                      ),
                    ),
                  ],
                ),
              ),
              floatingActionButton: OverlayTutorialHole(
                enabled: true,
                overlayTutorialEntry: OverlayTutorialRectEntry(
                  padding: const EdgeInsets.all(8.0),
                  radius: const Radius.circular(16.0),
                  overlayTutorialHints: [
                    OverlayTutorialWidgetHint(
                      builder: (context, entryRect) {
                        return Positioned(
                          top: entryRect.rRect!.top - 24.0,
                          left: entryRect.rRect!.left,
                          child: Text(
                            'Try this out',
                            style: textTheme.bodyText2!
                                .copyWith(color: tutorialColor),
                          ),
                        );
                      },
                    ),
                    OverlayTutorialWidgetHint(
                      position: (rect) => Offset(0, rect.center.dy),
                      builder: (context, entryRect) {
                        return SizedBox(
                          width: entryRect.rRect!.left,
                          child: Padding(
                            padding: const EdgeInsets.only(right: 8.0),
                            child: Row(
                              mainAxisAlignment: MainAxisAlignment.end,
                              children: [
                                Flexible(
                                  child: Text(
                                    'Click here to add counter',
                                    style: textTheme.bodyText2!
                                        .copyWith(color: tutorialColor),
                                  ),
                                ),
                              ],
                            ),
                          ),
                        );
                      },
                    ),
                  ],
                ),
                child: FloatingActionButton(
                  onPressed: _incrementCounter,
                  tooltip: 'Increment',
                  child: const Icon(Icons.add),
                ),
              ),
            ),
          ),
        ),
        if (_isTutorialEnabled)
          Align(
            alignment: Alignment.topRight,
            child: GestureDetector(
              onTap: () {
                _isTutorialEnabled = false;
                setState(() {});
              },
              child: Padding(
                padding: const EdgeInsets.all(8.0),
                child: Container(
                  padding: const EdgeInsets.all(16),
                  decoration: const BoxDecoration(
                    shape: BoxShape.circle,
                    color: Colors.black,
                  ),
                  child: const Icon(
                    Icons.close,
                    color: Colors.red,
                  ),
                ),
              ),
            ),
          ),
      ],
    ),
  );
}

```


# Hole Shape

- `OverlayTutorialRectEntry` draws a Rectangle shape
- `OverlayTutorialCircleEntry` draws a Circle shape
- `OverlayTutorialCustomShapeEntry` allows to draw a customized shape

# Disabling Widget

Disabling widget is not provided by this package out of the box. You can achieve this using one of the Flutter widget: `AbsorbPointer` or `IgnorePointer`. Make sure you do not wrap the `OverlayTutorialScope` with these widgets as this will disable the widget in `OverlayTutorialWidgetHint`.

# Screenshot

![](https://github.com/TabooSun/overlay_tutorial/blob/main/example/images/example_gif.gif?raw=true)

# API

Check [Dart Documentation](https://pub.dev/documentation/overlay_tutorial/latest/)

## License

[License](LICENSE).


CHANGELOG.md
## [3.0.0] - 12/5/2022

- Close [#40](https://github.com/TabooSun/overlay_tutorial/issues/40) - Flutter 3 compatibility. (Thanks [@daadu](https://github.com/daadu))

## [2.2.2] - 21/03/2022

- Fix broken screenshot in README.

## [2.2.1] - 21/03/2022

- Fix [#36](https://github.com/TabooSun/overlay_tutorial/issues/36).

## [2.2.0] - 20/03/2022

- Fix [#34](https://github.com/TabooSun/overlay_tutorial/issues/34). `OverlayTutorialScope` never calls `OverlayTutorialWidgetHint` builder of `OverlayTutorialCustomShapeEntry`.
  ### Breaking API Changes
    - `WidgetFromEntryBuilder` no longer provides `rect` & `rRect`. We merged them into `entryRect` of
      type `OverlayTutorialEntryRect`.
        - Migration steps:
      
          From
            ```dart
            (context, rect, rRect) {
              return Positioned(
                top: rRect.top,
                left: rRect.left,
                child: Text('Hello'),
              );
            }
            ``` 
          To
            ```dart
            (context, entryRect) {
              return Positioned(
                top: entryRect.rRect!.top,
                left: entryRect.rRect!.left,
                child: Text('Hello'),
              );
            }
            ```

## [2.1.1] - 9/12/2021

- Add test
- Fix for [flutter#91753](https://github.com/flutter/flutter/pull/91753) compatibility

## [2.1.0] - 6/11/2021

- Improve package performance.
    - Previous version tends to rebuild recursively non-stop which is fixed in this release.
- Add more example
    - Check the `example` project.
    - Due to community request, concrete example for disabling widget is added. Check `SimpleCounterTutorial` widget
      in `example` project.

## [2.0.4] - 19/6/2021

- Update CHANGELOG.md.

## [2.0.3] - 19/6/2021

- Patch [#23](https://github.com/TabooSun/overlay_tutorial/issues/23).
    - Fix widget rebuild too frequently
    - Fix FlutterError due to using setState when widget has already disposed

## [2.0.2] - 10/6/2021

- Patch [#20](https://github.com/TabooSun/overlay_tutorial/issues/20). (Thanks @Guang1234567)
    - Fix `OverlayTutorialHole.enabled` does not work.

## [2.0.1] - 4/5/2021

- Patch [#15](https://github.com/TabooSun/overlay_tutorial/issues/15).
    - Fix overlay is not rendering properly in web.

## [2.0.0] - 22/3/2021

- Complete [enhancement#10](https://github.com/TabooSun/overlay_tutorial/issues/10).
  ### Breaking API Changes
    1. ~~`OverlayTutorial`~~ is removed. Use `OverlayTutorialScope` instead.
    2. `OverlayTutorialScope` does not take all the `OverlayTutorialEntry`, pass these in new
       widget, `OverlayTutorialHole`.
    3. `OverlayTutorialHole` handles the `OverlayTutorialEntry`. Wrap `OverlayTutorialHole` with your widget that
       intends to make hole. Thus, you no longer need to provide any `GlobalKey`.
    4. `OverlayTutorialEntry` no longer takes `widgetKey` as argument.
    5. ~~`OverlayTutorialController`~~ is removed. Use the property in `OverlayTutorialScope` & `OverlayTutorialHole` to
       control the visibility of the overlay tutorial.

## [1.0.0] - 10/3/2021

- Support dart null-safety.

## [0.2.1] - 7/9/2020

- Allow package to be used on flutter 1.17.5

## [0.2.0] - 5/9/2020

- Fix Overlay Entry delay in production.

## [0.1.2] - 1/9/2020

- Hotfix - Overlay not showing on first time

## [0.1.1] - 1/9/2020

- Remove print statement
- Make `shapeBuilder` required in `OverlayTutorialCustomShapeEntry`
- Improve documentation.

## [0.1.0] - 29/8/2020

- Improve performance by preventing canvas from drawing continuously.

## [0.0.8] - 28/8/2020

* Bugfix & Improve documentation

## [0.0.7] - 28/8/2020

* Breaking API Changes: `OverlayTutorialEntry` is now an abstract class, use `OverlayTutorialRectEntry` instead.
* Introduce `OverlayTutorialCircleEntry` & `OverlayTutorialCustomShapeEntry` for drawing other shape.
* Change use of `addPersistentFrameCallback` to `addTimingsCallback` as the former will cause Flutter Inspector Select
  Widget mode to be malfunction.

## [0.0.6] - 27/8/2020

* Remove print statement

## [0.0.5] - 27/8/2020

* Show `overlayTutorialEntries` & `overlayChildren` only after `showOverlayTutorial` is called.

## [0.0.4] - 27/8/2020

* Handle SafeArea

## [0.0.3] - 26/8/2020

* Add Description to `pubspec.yaml`

## [0.0.2] - 26/8/2020

* Fix README image

## [0.0.1] - 26/8/2020

* Initial release
* Add OverlayTutorial widget


example/lib/views/simple_counter_tutorial.dart
/*
 * Copyright (C) TabooSun and contributors - All Rights Reserved
 * Written by TabooSun <taboosun1996@gmail.com>, 2021.
 */

import 'package:flutter/material.dart';
import 'package:overlay_tutorial/overlay_tutorial.dart';

class SimpleCounterTutorial extends StatefulWidget {
  final String title;

  const SimpleCounterTutorial({
    Key? key,
    required this.title,
  }) : super(key: key);

  @override
  _SimpleCounterTutorialState createState() => _SimpleCounterTutorialState();
}

class _SimpleCounterTutorialState extends State<SimpleCounterTutorial>
    with SingleTickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<Offset> _tweenAnimation;

  bool _isTutorialEnabled = true;

  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      duration: const Duration(milliseconds: 300),
      vsync: this,
    );

    _tweenAnimation = Tween<Offset>(
      begin: Offset.zero,
      end: const Offset(-1, 0),
    ).chain(CurveTween(curve: Curves.easeInOut)).animate(_animationController
      ..repeat(
        reverse: true,
      ));
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final textTheme = Theme.of(context).textTheme;
    const tutorialColor = Colors.yellow;
    return SafeArea(
      child: Stack(
        children: [
          OverlayTutorialScope(
            enabled: _isTutorialEnabled,
            overlayColor: Colors.blueAccent.withOpacity(.6),
            // Disable all the widgets. All the widgets are now non-interactive.
            child: AbsorbPointer(
              absorbing: _isTutorialEnabled,
              child: Scaffold(
                appBar: AppBar(
                  title: Text(widget.title),
                ),
                body: Center(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: <Widget>[
                      AnimatedBuilder(
                        animation: _tweenAnimation,
                        builder: (context, child) {
                          return SlideTransition(
                            position: _tweenAnimation,
                            child: child,
                          );
                        },
                        child: OverlayTutorialHole(
                          enabled: true,
                          overlayTutorialEntry: OverlayTutorialCustomShapeEntry(
                            shapeBuilder: (rect, path) {
                              path = Path.combine(
                                PathOperation.difference,
                                path,
                                Path()
                                  ..addOval(Rect.fromLTWH(
                                    rect.left - 16,
                                    rect.top,
                                    112,
                                    64,
                                  )),
                              );
                              return path;
                            },
                          ),
                          child: const Icon(
                            Icons.share,
                            size: 64,
                          ),
                        ),
                      ),
                      const SizedBox(height: 64),
                      const Text(
                        'You have pushed the button this many times:',
                      ),
                      OverlayTutorialHole(
                        enabled: true,
                        overlayTutorialEntry: OverlayTutorialRectEntry(
                          padding: const EdgeInsets.symmetric(horizontal: 8.0),
                          radius: const Radius.circular(16.0),
                          overlayTutorialHints: <OverlayTutorialWidgetHint>[
                            OverlayTutorialWidgetHint(
                              position: (rect) => Offset(0, rect.bottom),
                              builder: (context, entryRect) {
                                return SizedBox(
                                  width: MediaQuery.of(context).size.width,
                                  child: Center(
                                    child: Padding(
                                      padding: const EdgeInsets.all(8.0),
                                      child: Text(
                                        'Current Counter will be displayed here',
                                        style: textTheme.bodyText2!
                                            .copyWith(color: tutorialColor),
                                      ),
                                    ),
                                  ),
                                );
                              },
                            ),
                          ],
                        ),
                        child: Text(
                          '$_counter',
                          style: Theme.of(context).textTheme.headline4,
                        ),
                      ),
                    ],
                  ),
                ),
                floatingActionButton: OverlayTutorialHole(
                  enabled: true,
                  overlayTutorialEntry: OverlayTutorialRectEntry(
                    padding: const EdgeInsets.all(8.0),
                    radius: const Radius.circular(16.0),
                    overlayTutorialHints: [
                      OverlayTutorialWidgetHint(
                        builder: (context, entryRect) {
                          return Positioned(
                            top: entryRect.rRect!.top - 24.0,
                            left: entryRect.rRect!.left,
                            child: Text(
                              'Try this out',
                              style: textTheme.bodyText2!
                                  .copyWith(color: tutorialColor),
                            ),
                          );
                        },
                      ),
                      OverlayTutorialWidgetHint(
                        position: (rect) => Offset(0, rect.center.dy),
                        builder: (context, entryRect) {
                          return SizedBox(
                            width: entryRect.rRect!.left,
                            child: Padding(
                              padding: const EdgeInsets.only(right: 8.0),
                              child: Row(
                                mainAxisAlignment: MainAxisAlignment.end,
                                children: [
                                  Flexible(
                                    child: Text(
                                      'Click here to add counter',
                                      style: textTheme.bodyText2!
                                          .copyWith(color: tutorialColor),
                                    ),
                                  ),
                                ],
                              ),
                            ),
                          );
                        },
                      ),
                    ],
                  ),
                  child: FloatingActionButton(
                    onPressed: _incrementCounter,
                    tooltip: 'Increment',
                    child: const Icon(Icons.add),
                  ),
                ),
              ),
            ),
          ),
          if (_isTutorialEnabled)
            Align(
              alignment: Alignment.topRight,
              child: GestureDetector(
                onTap: () {
                  _isTutorialEnabled = false;
                  setState(() {});
                },
                child: Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: Container(
                    padding: const EdgeInsets.all(16),
                    decoration: const BoxDecoration(
                      shape: BoxShape.circle,
                      color: Colors.black,
                    ),
                    child: const Icon(
                      Icons.close,
                      color: Colors.red,
                    ),
                  ),
                ),
              ),
            ),
        ],
      ),
    );
  }
}
